sklearn/_display_estimator.py

```diff
@@ -0,0 +1,319 @@
+            if names is None:
+                names = (None, ) * len(estimators)
+            if name_details is None:
+                name_details = (None, ) * len(estimators)
+
+        self.names = names
+        self.name_details = name_details
+
+    def _sk_repr_html(self):
+        return self
+
+
+def _write_label_html(out, name, name_details,
+                      outer_class="sk-label-container",
+                      inner_class="sk-label",
+                      checked=False):
+    """Write labeled html with or without a dropdown with named details"""
+    out.write(f'<div class="{outer_class}">'
+              f'<div class="{inner_class} sk-toggleable">')
+
+    if name_details is not None:
+        checked_str = 'checked' if checked else ''
+        est_id = uuid.uuid4()
+        out.write(f'<input class="sk-toggleable__control sk-hidden--visually" '
+                  f'id="{est_id}" type="checkbox" {checked_str}>'
+                  f'<label class="sk-toggleable__label" for="{est_id}">'
+                  f'{name}</label>'
+                  f'<div class="sk-toggleable__content"><pre>{name_details}'
+                  f'</pre></div>')
+    else:
+        out.write(f'<label>{name}</label>')
+    out.write('</div></div>')  # outer_class inner_class
+
+
+def _get_visual_block(estimator):
+    """Generate information about how to display an estimator.
+    """
+    with suppress(AttributeError):
+        return estimator._sk_repr_html()
+
+    if isinstance(estimator, str):
+        return _VisualBlock('single', estimator,
+                            names=estimator, name_details=estimator)
+    elif estimator is None:
+        return _VisualBlock('single', estimator,
+                            names='None', name_details='None')
+
+    # check if estimator looks like a meta estimator wraps estimators
+    if hasattr(estimator, 'get_params'):
+        estimators = []
+        for key, value in estimator.get_params().items():
+            # Only look at the estimators in the first layer
+            if '__' not in key and hasattr(value, 'get_params'):
+                estimators.append(value)
+        if len(estimators):
+            return _VisualBlock('parallel', estimators, names=None)
+
+    return _VisualBlock('single', estimator,
+                        names=estimator.__class__.__name__,
+                        name_details=str(estimator))
+
+
+def _write_estimator_html(out, estimator, estimator_label,
+                          estimator_label_details, first_call=False):
+    """Write estimator to html in serial, parallel, or by itself (single).
+    """
+    if first_call:
+        est_block = _get_visual_block(estimator)
+    else:
+        # deeper calls will only show the changes
+        with config_context(print_changed_only=True):
+            est_block = _get_visual_block(estimator)
+
+    if est_block.kind == 'serial':
+        dashed_wrapped = first_call or est_block.dash_wrapped
```

Comment:
I think this and the next case could now be neatly refactored, but it's nice as it is